### PR TITLE
[wip] docker: 17.03.2-ce -> 17.06.0-ce

### PR DESCRIPTION
**Pull request by vdemeester**
Signed-off-by: Vincent Demeester <vincent@sbr.pm>

###### Motivation for this change

Release of `docker-ce` happened *and* the repository where the release happens has changed 😅 … 

It's still a `WIP` 

- because I'm getting an error I can't make sense of (and I don't see how to fix it)

```
/nix/store/bl48w7h24n6379fa200yv60n4pwynd9c-stdenv/setup: line 923: tracePhases: unbound variable
/nix/store/bl48w7h24n6379fa200yv60n4pwynd9c-stdenv/setup: line 125: showBuildStats: unbound variable
builder for ‘/nix/store/s67017nsynrhwipg9amhz1254z1245z0-docker-17.06.0-ce-rc2.drv’ failed with exit code 1
```

- I need to update `containerd` (probably elsewhere)

###### Things done

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/applications/virtualization/docker/default.nix [hunk 1]
```diff
@@ -140,31 +157,19 @@ rec {
...
-  docker_17_05 = dockerGen rec {
-    version = "17.05.0-ce";
-    rev = "90d35abf7b3535c1c319c872900fbd76374e521c"; # git commit
-    sha256 = "1m4fcawjj14qws57813wjxjwgnrfxgxnnzlj61csklp0s9dhg7df";
-    runcRev = "9c2d8d184e5da67c95d601382adf14862e4f2228";
-    runcSha256 = "131jv8f77pbdlx88ar0zjwdsp0a5v8kydaw0w0cl3i0j3622ydjl";
-    containerdRev = "9048e5e50717ea4497b757314bad98ea3763c145";
-    containerdSha256 = "1r9xhvzzh7md08nqb0rbp5d1rdr7jylb3da954d0267i0kh2iksa";
+  docker_17_06 = dockerGen rec {
+    version = "17.06.0-ce-rc2";
```
Comment (NeQuissimus) [inline pkgs/applications/virtualization/docker/default.nix]: We won't want RC2, right?

File: pkgs/applications/virtualization/docker/default.nix [hunk 2]
```diff
@@ -80,24 +80,41 @@ rec {
...
+      cd ./components/cli
+      # Mimic AUTO_GOPATH
+      mkdir -p .gopath/src/github.com/docker/
+	    ln -sf $PWD .gopath/src/github.com/docker/cli
+      export GOPATH="$PWD/.gopath:$GOPATH"
+      export GITCOMMIT="${rev}"
+      export VERSION="${version}"
+      source ./scripts/build/.variables
+      export CGO_ENABLED=1
+      go build -tags pkcs11 --ldflags "$LDFLAGS" github.com/docker/cli/cmd/docker
```
Comment (NeQuissimus) [inline pkgs/applications/virtualization/docker/default.nix]: I am glad you did this, I hate figuring out GO things :D

**Review by NeQuissimus — CHANGES_REQUESTED** [review]

File: pkgs/applications/virtualization/docker/default.nix (hunk 1 above)
Comment (vdemeester) [inline pkgs/applications/virtualization/docker/default.nix]: @NeQuissimus oh right, didn't update that yet.. was trying to figure out the errors before 😛

**Comment by NeQuissimus** [comment]
I got it, I think :)

I don't seem to be able to post commits into here, so here is the diff you need to apply, @vdemeester 

That weird error you were seeing was because `components/cli/scripts/build/.variables` does `set -eu`, which wrecks the rest of the build pretty good :)
Also, some of the files are in new locations, so I fixed up the installs.

```
diff --git a/pkgs/applications/virtualization/docker/default.nix b/pkgs/applications/virtualization/docker/default.nix
index 1fed24b14c..2eee6a4757 100644
--- a/pkgs/applications/virtualization/docker/default.nix
+++ b/pkgs/applications/virtualization/docker/default.nix
@@ -4,10 +4,6 @@
 , btrfs-progs, iptables, e2fsprogs, xz, utillinux, xfsprogs
 , procps
 }:
-
-# https://github.com/docker/docker/blob/master/project/PACKAGERS.md
-# https://github.com/docker/docker/blob/TAG/hack/dockerfile/binaries-commits
-
 with lib;
 
 rec {
@@ -78,25 +74,19 @@ rec {
       ++ optional (btrfs-progs == null) "exclude_graphdriver_btrfs"
       ++ optional (devicemapper == null) "exclude_graphdriver_devicemapper";
 
-    # systemd 230 no longer has libsystemd-journal as a separate entity from libsystemd
-    postPatch = ''
-      substituteInPlace ./components/engine/hack/make.sh                   --replace libsystemd-journal libsystemd
-      substituteInPlace ./components/engine/daemon/logger/journald/read.go --replace libsystemd-journal libsystemd
-    '';
-
     buildPhase = ''
       # build engine
-      patchShebangs .
       cd ./components/engine
       export AUTO_GOPATH=1
       export DOCKER_GITCOMMIT="${rev}"
       ./hack/make.sh dynbinary
       cd -
+
       # build cli
       cd ./components/cli
       # Mimic AUTO_GOPATH
       mkdir -p .gopath/src/github.com/docker/
-	    ln -sf $PWD .gopath/src/github.com/docker/cli
+      ln -sf $PWD .gopath/src/github.com/docker/cli
       export GOPATH="$PWD/.gopath:$GOPATH"
       export GITCOMMIT="${rev}"
       export VERSION="${version}"
@@ -104,7 +94,14 @@ rec {
       export CGO_ENABLED=1
       go build -tags pkcs11 --ldflags "$LDFLAGS" github.com/docker/cli/cmd/docker
       cd -
-      echo foo
+    '';
+
+    # systemd 230 no longer has libsystemd-journal as a separate entity from libsystemd
+    patchPhase = ''
+      patchShebangs .
+      substituteInPlace ./components/engine/hack/make.sh                   --replace libsystemd-journal libsystemd
+      substituteInPlace ./components/engine/daemon/logger/journald/read.go --replace libsystemd-journal libsystemd
+      substituteInPlace ./components/cli/scripts/build/.variables --replace "set -eu" ""
     '';
 
     outputs = ["out" "man"];
@@ -112,8 +109,7 @@ rec {
     extraPath = makeBinPath [ iproute iptables e2fsprogs xz xfsprogs procps utillinux ];
 
     installPhase = ''
-      echo bar
-      install -Dm755 ./components/engine/bundles/${version}/dynbinary-client/docker-${version} $out/libexec/docker/docker
+      install -Dm755 ./components/cli/docker $out/libexec/docker/docker
       install -Dm755 ./components/engine/bundles/${version}/dynbinary-daemon/dockerd-${version} $out/libexec/docker/dockerd
       makeWrapper $out/libexec/docker/docker $out/bin/docker \
         --prefix PATH : "$out/libexec/docker:$extraPath"
@@ -131,15 +127,15 @@ rec {
       install -Dm644 ./components/engine/contrib/init/systemd/docker.service $out/etc/systemd/system/docker.service
 
       # completion (cli)
-      install -Dm644 ./contrib/completion/bash/docker $out/share/bash-completion/completions/docker
-      install -Dm644 ./contrib/completion/fish/docker.fish $out/share/fish/vendor_completions.d/docker.fish
-      install -Dm644 ./contrib/completion/zsh/_docker $out/share/zsh/site-functions/_docker
+      install -Dm644 ./components/cli/contrib/completion/bash/docker $out/share/bash-completion/completions/docker
+      install -Dm644 ./components/cli/contrib/completion/fish/docker.fish $out/share/fish/vendor_completions.d/docker.fish
+      install -Dm644 ./components/cli/contrib/completion/zsh/_docker $out/share/zsh/site-functions/_docker
 
       # Include contributed man pages (cli)
-      man/md2man-all.sh -q
+      ./components/cli/man/md2man-all.sh -q
       manRoot="$man/share/man"
       mkdir -p "$manRoot"
-      for manDir in man/man?; do
+      for manDir in ./components/cli/man/man?; do
         manBase="$(basename "$manDir")" # "man1"
         for manFile in "$manDir"/*; do
           manName="$(basename "$manFile")" # "docker-build.1"
@@ -162,10 +158,12 @@ rec {
     };
   };
 
+  # https://github.com/docker/docker-ce/blob/v${version}/components/engine/hack/dockerfile/binaries-commits
+
   docker_17_06 = dockerGen rec {
-    version = "17.06.0-ce-rc2";
-    rev = "402dd4a9ea3802e45718f871b00efe52caae5108"; # git commit
-    sha256 = "05x7bzjwmnab4bnnbcqj7rrxhv16a0ys71xwskl5f7adm3ic696g";
+    version = "17.06.0-ce";
+    rev = "02c1d876176546b5f069dae758d6a7d2ead6bd48"; # git commit
+    sha256 = "0wrg4ygcq4c7f2bwa7pgc7y33idg0hijavx40588jaglz4k8sqpm";
     runcRev = "992a5be178a62e026f4069f443c6164912adbf09";
     runcSha256 = "0ylkbn5rprw5cgxazvrwj7balikpfm8vlybwdbfpwnsqk3gc6p8k";
     containerdRev = "cfb82a876ecc11b5ca0977d1733adbe58599088a";
```

**Comment by vdemeester** [comment]
@NeQuissimus thanks a lot ! Updated and it is working 🌥 😎 👍. Also updated the man generation stuff.

**Pull request merged**